csharp/extractor/Semmle.Extraction/Context.cs

```diff
@@ -52,6 +54,10 @@ protected void DefineLabel(IEntity entity)
                 finally
                 {
                     writingLabel = false;
+                    if (labelQueue.Any())
```

Comment:
Should this be `while` loop rather than an `if` so that we write out _all_ the labels from the queue?

Reply:
`DefineLabel` is recursive, so a `while` loop is not needed.

Reply:
This recursion might result in strange behavior: the first item (`entity A`) in the queue is dequeued, and if in the meantime `DefineLabel` is being called on another entity, then `entity A` might be enqueued again, which means it goes to the end of the queue, so the queue doesn't give any guarantee on the ordering.

Reply:
The order is not super important, what matters is that we populate the labels as early as possible. For example, for a method like
```csharp
IEnumerable<int> M(string s) { }
```
we will (simplified) generate the following TRAP:

```
#0 = {#1} M({#2} s)
#1 = {#3}<{#4}>
#2 = "string"
#3 = "IEnumerable<>"
#4 = "int"
```

so we are still using labels before their definition, but the definition will follow shortly after, as opposed to the old implementation.